include/nvexec/stream/common.cuh

```diff
@@ -386,34 +442,65 @@ namespace nvexec {
         context_state_t context_state_;
         void* temp_storage_{nullptr};
         outer_receiver_t rcvr_;
-        cudaError_t status_{cudaSuccess};
-        std::optional<cudaStream_t> own_stream_{};
-        bool defer_stream_destruction_{false};
+        stream_provider_t stream_provider_;
 
         __t(outer_receiver_t rcvr, context_state_t context_state, bool defer_stream_destruction)
           : context_state_(context_state)
           , rcvr_(rcvr)
-          , defer_stream_destruction_(defer_stream_destruction) {
-          if constexpr (!borrows_stream) {
-            std::tie(own_stream_, status_) = create_stream_with_priority(context_state_.priority_);
-          }
+          , stream_provider_(defer_stream_destruction, borrows_stream, context_state.priority_) {
         }
 
-        cudaStream_t get_stream() const {
-          cudaStream_t stream{};
+        stream_provider_t* get_stream_provider() const {
+          stream_provider_t *stream_provider{};
 
           if constexpr (borrows_stream) {
             const outer_env_t& env = get_env(rcvr_);
-            stream = ::nvexec::STDEXEC_STREAM_DETAIL_NS::get_stream(env);
+            stream_provider = ::nvexec::STDEXEC_STREAM_DETAIL_NS::get_stream_provider(env);
           } else {
-            stream = *own_stream_;
+            stream_provider = &const_cast<stream_provider_t&>(stream_provider_);
           }
 
-          return stream;
+          return stream_provider;
+        }
+
+        cudaStream_t get_stream() const {
+          return get_stream_provider()->own_stream_.value();
+        }
+
+        template <class T>
+        void defer_temp_storage_destruction(T *ptr) {
+          assert(ptr == this->temp_storage_);
+
+          if constexpr (!std::is_trivially_destructible_v<T>) {
+            temp_storage_ = nullptr; // defer deallocation to the stream provider
+            stream_provider_t *stream_provider = get_stream_provider();
+            std::pmr::memory_resource* managed_resource = context_state_.managed_resource_;
+
+            // Stream is destroyed when the last object is buried, so it's safe to use it here
+            cudaStream_t stream = stream_provider->own_stream_.value();
+            stream_provider->bury([ptr, stream, managed_resource] {
+              std::int32_t device_id = cudaInvalidDeviceId;
+
+              cudaMemRangeGetAttribute(
+                &device_id, 4, cudaMemRangeAttributeLastPrefetchLocation, ptr, sizeof(T));
+
+              if (cudaCpuDeviceId == device_id) {
+                ptr->~T();
+              } else {
+                destructor_kernel<<<1, 1, 0, stream>>>(ptr);
+
+                // TODO Bury all the memory associated with the stream provider and then 
+                //      deallocate the memory
+                cudaStreamSynchronize(stream); 
+              }
+
+              managed_resource->deallocate(ptr, sizeof(T));
+            });
+          }
```

Comment:
The purist in me wants to see the registration of the destroy/deallocate action to happen as part of the allocate/construct action, rather than as a separate step on a separate data structure. Is that possible?

Reply:
Registration of deallocation has to happen at a particular place (before propagation of the completion signal and after some kernel) in a particular condition (no CUDA errors). I don't think it can happen as part of allocation/construction action without some additional memory traffic.